torchtune/modules/peft/mora.py

```diff
@@ -0,0 +1,194 @@
+        _mora_init_params(self.lora_a)
+        # _mora_init_params(self.lora_b)
+
+    def _create_weight_and_bias(self):
+        """
+        Creates a linear weight and bias tensor, using NF4 dtype if we're quantizing
+        (indicated via quantize_base=True).
+        """
+        in_dim, out_dim, use_bias = self.in_dim, self.out_dim, self.use_bias
+        linear = nn.Linear(in_features=in_dim, out_features=out_dim, bias=use_bias)
+        weight = linear.weight if not self._quantize_base else to_nf4(linear.weight)
+        bias = None
+        if self.use_bias:
+            if self._quantize_base:
+                raise NotImplementedError(
+                    "Quantized MoRALinear does not support bias at the moment."
+                )
+            bias = linear.bias
+        return weight, bias
+
+    def adapter_params(self) -> List[str]:
+        """
+        Return lora_a.weight as adapter params.
+        If bias is enabled, also return lora_a.bias
+        """
+        # NOTE: this function has to be updated if the names of "lora_a" and "lora_b"
+        # in this module change.
+        adapter_params = [
+            "lora_a.weight",
+        ]  # "lora_b.weight"]
+        return adapter_params
+
+    def precompute_freqs(self, r: int, base: int = 10000):
+        in_f = self.in_dim
+        rb1 = in_f // r if in_f % r == 0 else in_f // r + 1
+        inv_freq = 1.0 / (base ** (torch.arange(0, r, 2).float() / r))
+        t = torch.arange(rb1)
+        freqs = torch.outer(t, inv_freq)
+        emb = torch.cat((freqs, freqs), dim=-1)
+        return emb.cos().unsqueeze(0), emb.sin().unsqueeze(0)
+
+    def _apply_mora(self, x):
+        """
+        Taken from the paper author's github (Apache 2 License):
+        https://github.com/kongds/MoRA/blob/0ff64b144e60b54fe7c0ff7b4e76c99c949e923d/peft-mora/src/peft/tuners/lora/layer.py#L229
+        """
+        r = self.rank
+        in_f, out_f = self.in_dim, self.out_dim
+        sum_inter = in_f // r
+        rb1 = in_f // r if in_f % r == 0 else in_f // r + 1
+        if in_f % r != 0:
+            pad_size = r - in_f % r
+            x = torch.cat([x, x[..., :pad_size]], dim=-1)
+            sum_inter += 1
+        in_x = x.view(*x.shape[:-1], sum_inter, r)
+        if not hasattr(self, "cos") and not hasattr(self, "sin"):
+            self.cos, self.sin = self.precompute_freqs(r)
+        rh_in_x = torch.cat((-in_x[..., r // 2 :], in_x[..., : r // 2]), dim=-1)
+        in_x = in_x * self.cos + rh_in_x * self.sin
+        out_x = self.lora_a(in_x.to(self.lora_a.weight.dtype))
+        out_x = out_x.view(*x.shape[:-1], -1)[..., :out_f]
+        if out_x.shape[-1] < out_f:
+            repeat_time = out_f // out_x.shape[-1]
+            if out_f % out_x.shape[-1] != 0:
+                repeat_time += 1
+            out_x = torch.cat([out_x] * repeat_time, dim=-1)[..., :out_f]
```

Comment:
Can we clean this up a bit? Personally I find it a bit hard to parse what's happening here. Imo it's best to just write from scratch with shape comments and explicit variable names along the way